File: basic-service-invocation/microservices/svc/email/handler/email.go
package handler

import (
	"context"

	pb "github.com/kzmake/examples/basic-service-invocation/gen/go/email/v1"
)

type email struct {
	pb.UnimplementedEmailServiceServer
}

var _ pb.EmailServiceServer = new(email)

func NewEmail() pb.EmailServiceServer {
	return &email{}
}

func (h *email) Create(ctx context.Context, req *pb.CreateRequest) (*pb.CreateResponse, error) {
	email := &pb.Email{
		Address: req.GetUsername() + "@example.com",
	}

	// do something

	return &pb.CreateResponse{Email: email}, nil
}

File: basic-service-invocation-py/microservices/svc/email/handler/email.py
from basic.email.v1.email_grpc import EmailServiceBase
from basic.email.v1.email_pb2 import CreateRequest, CreateResponse, Email
from grpclib.server import Stream


class EmailService(EmailServiceBase):
    async def Create(self, stream: Stream[CreateRequest, CreateResponse]) -> None:
        req = await stream.recv_message()
        assert req is not None

        email = Email(address=f"{req.username}@example.com")

        # do something

        res = CreateResponse(email=email)
        await stream.send_message(res)
